Add workspace alert to sprite lab when sprites reach max

fisher-alice opened the pull request:
- This is a follow-up task to [this PR](https://github.com/code-dot-org/code-dot-org/pull/47738) which limits the number of sprites created in Sprite Lab to prevent the browser from freezing. Once this PR is approved, it will be merged with https://github.com/code-dot-org/code-dot-org/pull/47953 and then both parts will be merged to staging together.
- This PR displays a workspace alert warning when the user creates more than MAX_NUM_SPRITE_WARNING (999) sprites either during preview mode (auto-run) or when the program is running. The workspace alert disappears if the number of sprites created is MAX_NUM_SPRITE_WARNING or less during preview mode or when the program is reset.
- The MAX_NUM_SPRITES is set to 1000. When number of sprites is MAX_NUM_SPRITES, functions that create sprites return early which drastically improves performance.
- This PR builds on [this update](https://github.com/code-dot-org/code-dot-org/pull/47815) which added the workspace alert to the CodeWorkspace component that can be used in both App lab and Sprite lab.

## After update videos

- This video shows the workspace alert warning being displayed and closed for the function addSprite:

https://user-images.githubusercontent.com/107423305/188748396-a7688ac8-882c-47b0-a9ca-ee4054aff7c0.mp4

- This video shows the workspace alert warning being displayed and closed for the function addNumSprites:

https://user-images.githubusercontent.com/107423305/188748415-f17179aa-9cf0-4e77-9cbc-62087f454949.mp4


- This video shows the workspace alert warning being displayed and closed for the function addBursts:

https://user-images.githubusercontent.com/107423305/188748472-ef898a26-4a16-497b-944a-49e34b07e721.mp4


- This video shows the workspace alert warning being displayed and closed for two different adding sprite functions. It also shows the result when the number of iterations is very large.

https://user-images.githubusercontent.com/107423305/188748492-6ef93dfb-4d16-46a8-8e28-7e8a431a85d0.mp4


## Links
- [jira ticket - [Sprite Lab] Display workspace alert when user reaches limit](https://codedotorg.atlassian.net/browse/STAR-2376)
- [jira ticket - Prototype: How can we allow users to block the auto-run functionality in Sprite Lab](https://codedotorg.atlassian.net/browse/STAR-2323)

## Testing story
- In CoreLibraryTest.js added a unit test to check that number of sprites are limited when addSprite is called within a loop with large number of iterations. Also added a unit test to ensure a workspace alert is displayed after MAX_NUM_SPRITE_WARNING sprites has been created.

## Follow-up work
In Sprite Lab, there are additional ways to cause the browser to freeze
- A user can add behaviors to sprites. The following program will not cause the browser to freeze with this PR update:

<img width="482" alt="Screen Shot 2022-08-31 at 12 22 03 PM" src="https://user-images.githubusercontent.com/107423305/187740717-d6d7d072-7dff-4c70-a19f-12d9a106db1c.png">

Although it doesn't necessarily make sense to do so, If a user adds a sprite behavior to a loop with a large number of iterations (>10,000), this will cause the browser to freeze. 

<img width="523" alt="Screen Shot 2022-08-31 at 12 22 42 PM" src="https://user-images.githubusercontent.com/107423305/187740820-def9d0e6-8992-4f08-a9ad-5da79bad8aa7.png">

- Using the print function within a loop will also cause the browser to freeze for large number of iterations (>500).

Both of these issues should be addressed:
- [jira - Large number of print statements cause browser freezing](https://codedotorg.atlassian.net/browse/STAR-2395)
- [jira - Large number of sprite behaviors cause browser freezing](https://codedotorg.atlassian.net/browse/STAR-2396) 

## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

sanchitmalhotra126 commented on apps/i18n/common/en_us.json:
```diff
@@ -1959,6 +1959,7 @@
   "specialAnnouncementHeadingPostHoc2019": "Go beyond an Hour of Code",
   "specialAnnouncementDescriptionPostHoc2019": "An hour is just the beginning. We offer comprehensive curriculum for grades K-12, professional learning programs including local workshops, and resources to help you and your school get started.",
   "sprite": "sprite",
+  "spriteLimitExceeded": "The maximum possible number of sprites in Sprite Lab is 1000. No additional sprites can be created.",
```
It would be good to make the 1000 in this string configurable in case we decide to alter the limit later. If you include something like `{limit}` instead of the 1000 (ex. `...number of sprites in Spite Lab is {limit}. No additional...`), then when referencing this string, you can pass in an object specifying the value of "limit", ex. `msg.spriteLimitExceeded({limit: MAX_NUM_SPRITES})`

fisher-alice replied:
That's really cool! Thanks - will update.

sanchitmalhotra126 commented on apps/src/p5lab/redux/textConsole.js:
```diff
@@ -7,8 +7,8 @@ export default function textConsole(state, action) {
-        // 1000 items is an arbitrary limit. Change as needed.
-        ...state.slice(0, 999),
+        // 1000 is an arbitrary limit - the last 1000 text console statements will be displayed
+        ...state.slice(-999),
```
Should we store this 999 in constants as well? (or store it as a 1000 and convert it to -999 here so the actual limit value is clearer)

fisher-alice replied:
Yes - this makes sense! I updated in the prior PR that first made this change. https://github.com/code-dot-org/code-dot-org/pull/47738

sanchitmalhotra126 commented on apps/src/p5lab/spritelab/CoreLibrary.js:
```diff
@@ -345,19 +352,56 @@ export default class CoreLibrary {
+  // a workspace alert warning is displayed.
+  // (MAX_NUM_SPRITES + 1) is the actual maximum number of sprites possible
+  // and once this number is reached, this function returns true
+  checkReachSpriteLimit() {
```
nit: maybe `hasReachedSpriteLimit` or `reachedSpriteLimit` since this returns a boolean?

fisher-alice replied:
Makes sense! Will use reachedSpriteLimit.

sanchitmalhotra126 commented on apps/test/unit/p5lab/spritelab/CoreLibraryTest.js:
```diff
@@ -169,6 +176,36 @@ describe('SpriteLab Core Library', () => {
 
+  describe('Sprite limit', () => {
+    // The number of sprites created is limited by (MAX_NUM_SPRITES + 1)
+    it('caps at (MAX_NUM_SPRITES + 1) or 1001 sprites', () => {
```
nit: if you want to make this test future-proof in case MAX_NUM_SPRITES changes, you can convert the test name to a template string and replace the 1001 with ${MAX_NUM_SPRITES + 1} 🙂

jmkulwik commented:
It looks like this branch is based on `sprite_lab_limit_sprites/texts` from https://github.com/code-dot-org/code-dot-org/pull/47738/commits.
If that's the case, go ahead and update the base branch on the pull request to `sprite_lab_limit_sprites/texts` for ease of reviewing. You can do this by clicking the "edit" button to the right of the PR title and changing the base branch from staging. That also makes it clearer to future developers what code is changed in this PR specifically

sanchitmalhotra126 commented on apps/test/unit/p5lab/spritelab/CoreLibraryTest.js:
```diff
@@ -169,6 +176,36 @@ describe('SpriteLab Core Library', () => {
+  describe('Sprite limit', () => {
+    // The number of sprites created is limited by (MAX_NUM_SPRITES + 1)
+    it('caps at (MAX_NUM_SPRITES + 1) or 1001 sprites', () => {
+      for (var i = 0; i < 50000; i++) {
```
nit: `let`

sanchitmalhotra126 commented on apps/test/unit/p5lab/spritelab/CoreLibraryTest.js:
```diff
@@ -169,6 +176,36 @@ describe('SpriteLab Core Library', () => {
+      stub(redux, 'getStore').returns({
+        dispatch: stubbedDispatch
+      });
+      for (var i = 0; i < 1001; i++) {
```
nit: could use MAX_NUM_SPRITES + 1 in place of 1001 here

sanchitmalhotra126 reviewed (commented):
Nice! Just a few comments

fisher-alice commented:
> It looks like this branch is based on `sprite_lab_limit_sprites/texts` from https://github.com/code-dot-org/code-dot-org/pull/47738/commits. If that's the case, go ahead and update the base branch on the pull request to `sprite_lab_limit_sprites/texts` for ease of reviewing. You can do this by clicking the "edit" button to the right of the PR title and changing the base branch from staging. That also makes it clearer to future developers what code is changed in this PR specifically

Thanks Jessie! I merged the `sprite_lab_limit_sprites/texts` with staging to create this branch since I wanted to also access the updates I made from https://github.com/code-dot-org/code-dot-org/pull/47815 so if I change the base branch to `sprite_lab_limit_sprites/texts`, I get many more changes than the ones I've made. 

I merged the prior PR so now the diff should be more clear. Thanks!

pablo-code-org commented on apps/src/p5lab/spritelab/CoreLibrary.js:
```diff
@@ -345,19 +352,56 @@ export default class CoreLibrary {
+  // and once this number is reached, this function returns true
+  reachedSpriteLimit() {
+    const numSprites = this.getNumberOfSprites();
+    if (numSprites >= MAX_NUM_SPRITES + 1) {
```
small nit :) 

numSprites > MAX_NUM_SPRITES 

In my head is more human readable and matches your method definition without an operation going through the devs head.

pablo-code-org replied:
sorry, my bad, 

I get confused with the definition of Max_NUM_SPRITES which is not the real max 
`(MAX_NUM_SPRITES + 1) is the actual maximum number of sprites possible`

pablo-code-org replied:
Can a new variable be created 

MAX_NUM_SPRITES = 1001;
MAX_NUM_SPRITE_WARNING = MAX_NUMBER_SPRITES - 1;

And then it would avoid confusions when reading the code? We could even add some comments around the second variable to understand more about it. I can see someone forgetting a +1 somewhere and introducing a bug

fisher-alice replied:
I LOVE this idea of using two constants. I was struggling with which one to use, but much better to use 2 different constants! And I like the name MAX_NUM_SPRITE_WARNING - will use. Thanks so much!

jmkulwik commented on apps/src/p5lab/spritelab/CoreLibrary.js:
```diff
@@ -366,16 +369,38 @@ export default class CoreLibrary {
+    if (numSprites > MAX_NUM_SPRITE_WARNING) {
+      return true;
+    }
+    if (numSprites === MAX_NUM_SPRITE_WARNING) {
```
For my own understanding: Do we have the `=== MAX_NUM_SPRITE_WARNING` so the warning gets displayed only once per time the user goes over the max number of sprites? i.e. if the user puts 1005, the warning is displayed once instead of 5x?

fisher-alice replied:
Yes, that's correct.

jmkulwik replied:
(i.e. as opposed to using `>= MAX_NUM_SPRITES`)

jmkulwik replied:
(oops I had the page cached and didn't see your response. Ignore my second comment)

jmkulwik commented on apps/src/p5lab/spritelab/CoreLibrary.js:
```diff
@@ -366,16 +369,38 @@ export default class CoreLibrary {
+  // When the total number of sprites reaches MAX_NUM_SPRITE_WARNING,
+  // a workspace alert warning is displayed.
+  // MAX_NUM_SPRITE_WARNING = MAX_NUM_SPRITES - 1
+  reachedSpriteLimit() {
```
This function reads like a boolean function, but has a side effect of triggering a warning on the code editor. This can lead to confusion when maintaining the code because someone in the future might copy what looks like a simple boolean function and get unintended side effects.

This can be fixed by splitting this into two functions. Something like `reachedSpriteLimit` and `warnIfAtSpriteLimit`.

fisher-alice replied:
Thanks Jessie! This is great feedback.
I updated reachedSpriteLimit so it just checks if the total number of sprites >= MAX_NUM_SPRITES.
Then added a warnIfAtSpriteLimit to dispatch a warning if the total number of sprites == SPRITE_LIMIT_WARNING. I renamed the constant to reflect its purpose more clearly. 
I think that separating these 2 functions also clarifies the purpose of the 2 constants as well to address your comment below about the use of 2 constants instead of 1. Lmk what you think of this update.

I added a couple more unit tests to check to make sure the alert is dispatched correctly - a test to check warning is dispatched only once even when the addSprites function is called additional times past the SPRITE_LIMIT_WARNING. And then a test to make sure the alert is not dispatched at the edge case.

Thanks!

jmkulwik commented on apps/src/p5lab/spritelab/constants.js:
```diff
@@ -100,6 +100,9 @@ export const exampleSprites = {
+// The user will receive a workspace alert at MAX_NUM_SPRITE_WARNING. But the actual cap
+// is MAX_NUM_SPRITES
+export const MAX_NUM_SPRITES = 1001;
+export const MAX_NUM_SPRITE_WARNING = MAX_NUM_SPRITES - 1;
```
I'm concerned that the MAX_NUM_SPRITES and MAX_NUM_SPRITE_WARNING could lead to future off-by-one errors that are difficult to catch. A future code maintainer would need to be very careful that they're using the right variable in the right place. Since 1000 is a somewhat arbitrary limit (i.e. choosing 1010 or 995 would lead to no real difference in user experience), just using `MAX_NUM_SPRITES = 1000` and limiting the user if `spritesRequested > MAX_NUM_SPRITES` and displaying the warning if `spritesRequested === MAX_NUM_SPRITES` seems very reasonable to me. Then there's only one variable that needs to be used everywhere and it's a little more clear to the reader what is happening in each place. The user _could_ create 1001 sprites, but the intended functionality still works. However, with the two competing variables, I worry that a future code maintainer may accidentally break the functionality and allow users to create as many sprites as they want.

I'm curious to hear others thoughts, however. This is how I would approach it, but I could see an argument in a different direction.

fisher-alice replied:
Yes - I would also like to hear other folks's thoughts as well!
I renamed the MAX_NUM_SPRITE_WARNING to SPRITE_LIMIT_WARNING and made other updates in its use (created a 2nd function) and tests. More details in my reply to comment above. Thanks!

breville replied:
I agree with @jmkulwik that it seems a bit odd to export two numbers, one of which is always one less than the other, when you really only need to export one of them, and the business logic elsewhere can handle the warning when you are exactly one before.  However, there is one argument for keeping them as separate, and that would be if we might want to adjust this to warn at a notably different threshold than the maximum.  If that's never going to happen, though, then it seems more robust to just export one magic number.

fisher-alice replied:
Thanks @breville ! This makes and I'm definitely open to refactor.

jmkulwik commented on apps/test/unit/p5lab/spritelab/CoreLibraryTest.js:
```diff
@@ -169,6 +179,38 @@ describe('SpriteLab Core Library', () => {
+      for (let i = 0; i < MAX_NUM_SPRITES; i++) {
+        coreLibrary.addSprite();
+      }
+      expect(stubbedDispatch).to.have.been.calledWith(
```
NIT: should `calledWith` here be `calledOnceWith`? ([sinon-chai cheat sheet](https://www.chaijs.com/plugins/sinon-chai/) if it's helpful)

fisher-alice replied:
Thanks! Made the update.

fisher-alice replied:
Also added 2 more unit tests.

mikeharv commented:
> > It looks like this branch is based on `sprite_lab_limit_sprites/texts` from https://github.com/code-dot-org/code-dot-org/pull/47738/commits. If that's the case, go ahead and update the base branch on the pull request to `sprite_lab_limit_sprites/texts` for ease of reviewing. You can do this by clicking the "edit" button to the right of the PR title and changing the base branch from staging. That also makes it clearer to future developers what code is changed in this PR specifically
> 
> Thanks Jessie! I merged the `sprite_lab_limit_sprites/texts` with staging since I wanted to also access the updates I made from #47815 so if I change the base branch to `sprite_lab_limit_sprites/texts`, I get many more changes than the ones I've made.
> 
> I merged the prior PR so now the diff should be more clear. Thanks!

Hey Alice, I think we had talked about not merging https://github.com/code-dot-org/code-dot-org/pull/47738/ until this one was ready to go, that way we don't release a feature that only has half of what we want. I'd suggest reverting https://github.com/code-dot-org/code-dot-org/pull/47738/ and merging that revert ASAP. 

Once that's done, you can set up another PR to "revert the revert". Jessie's suggestion was to update the base branch at the top of this page to point to the original changes, but still unmerged. (See screenshot)

<img width="1476" alt="image" src="https://user-images.githubusercontent.com/43474485/188049268-b339f008-67f4-4974-bc62-553a79b14944.png">

That should just make it easier to compare the changes here without also having to make sense of the changes from the original PR. Once this one is approved, you'd want to merge this one _first_. 

I hope this is somewhat clear, and I apologize if it's not! Thanks for all the great work.

fisher-alice commented:
> > > It looks like this branch is based on `sprite_lab_limit_sprites/texts` from https://github.com/code-dot-org/code-dot-org/pull/47738/commits. If that's the case, go ahead and update the base branch on the pull request to `sprite_lab_limit_sprites/texts` for ease of reviewing. You can do this by clicking the "edit" button to the right of the PR title and changing the base branch from staging. That also makes it clearer to future developers what code is changed in this PR specifically
> > 
> > 
> > Thanks Jessie! I merged the `sprite_lab_limit_sprites/texts` with staging since I wanted to also access the updates I made from #47815 so if I change the base branch to `sprite_lab_limit_sprites/texts`, I get many more changes than the ones I've made.
> > I merged the prior PR so now the diff should be more clear. Thanks!
> 
> Hey Alice, I think we had talked about not merging #47738 until this one was ready to go, that way we don't release a feature that only has half of what we want. I'd suggest reverting #47738 and merging that revert ASAP.
> 
> Once that's done, you can set up another PR to "revert the revert". Jessie's suggestion was to update the base branch at the top of this page to point to the original changes, but still unmerged. (See screenshot)
> 
> <img alt="image" width="1476" src="https://user-images.githubusercontent.com/43474485/188049268-b339f008-67f4-4974-bc62-553a79b14944.png">
> 
> That should just make it easier to compare the changes here without also having to make sense of the changes from the original PR. Once this one is approved, you'd want to merge this one _first_.
> 
> I hope this is somewhat clear, and I apologize if it's not! Thanks for all the great work.

Thanks Mike! I did try Jessie's suggestion to update the base branch at first. However, since I merged this branch with the latest version of staging to pull in my changes from https://github.com/code-dot-org/code-dot-org/pull/47815, the diff included all the other changes made to staging since I last pulled (yikes!). So was attempting to make it easier for Jessie to review this PR compared to the #47738.
She since was able to review so I reverted #47738 and plan to merge both PRs at the same time. Thanks so much!
Update: I reverted and differences seem to be maintained correctly :)

sanchitmalhotra126 commented on apps/src/p5lab/spritelab/CoreLibrary.js:
```diff
@@ -366,12 +369,41 @@ export default class CoreLibrary {
   }
 
+  reachedSpriteMax() {
+    const numSprites = this.getNumberOfSprites();
+    return numSprites >= MAX_NUM_SPRITES;
```
nit: you could probably just inline this? `return this.getNumberOfSprites() >= MAX_NUM_SPRITES`

fisher-alice replied:
Thanks! Made the update.

sanchitmalhotra126 commented on apps/src/p5lab/spritelab/CoreLibrary.js:
```diff
@@ -366,12 +369,41 @@ export default class CoreLibrary {
+  // a workspace alert warning is displayed
+  // Note that SPRITE_LIMIT_WARNING = MAX_NUM_SPRITES - 1
+  warnIfAtSpriteLimit() {
+    const numSprites = this.getNumberOfSprites();
+    if (numSprites === SPRITE_LIMIT_WARNING) {
```
nit: same here as well (`if (this.getNumberOfSprites() === SPRITE_LIMIT_WARNING)`)

fisher-alice replied:
Thanks! Made the update.

sanchitmalhotra126 commented on apps/src/p5lab/spritelab/CoreLibrary.js:
```diff
@@ -366,12 +369,41 @@ export default class CoreLibrary {
    * @returns {Number} A unique id to reference the sprite.
    */
   addSprite(opts) {
+    // this function returns early if total number of sprites equals MAX_NUM_SPRITES
+    if (this.reachedSpriteMax()) {
+      return;
+    }
+    // this function dispatches a workspace alert if the total number of sprites
+    // equals SPRITE_LIMIT_WARNING
+    this.warnIfAtSpriteLimit();
```
Minor nit, but since this set of code is repeated a few times, I don't think it needs to be commented this heavily in each case - more potential for one comment to get out of sync with others which could lead to confusion down the road. As long as the original functions are well commented (which `warnIfAtSpriteLimit` seems to be; may want to add a quick comment to `reachedSpriteMax`, though it is pretty self-explanatory 🙂) I don't think you need to include too many comments here.

fisher-alice replied:
Thanks - makes sense to remove repetitive comments. Made the updates!

sanchitmalhotra126 reviewed (approved):
Looks good to me! Thanks for working through the branches and updating the PR! Just a few minor nits.

mikeharv reviewed (approved):
Fantastic work Alice! Thanks for all of the persistence on this one. :)

jmkulwik commented on apps/src/p5lab/spritelab/constants.js:
```diff
@@ -100,6 +100,9 @@ export const exampleSprites = {
+// The user will receive a workspace alert at SPRITE_LIMIT_WARNING. But the actual cap
+// is MAX_NUM_SPRITES
 export const MAX_NUM_SPRITES = 1000;
+export const SPRITE_LIMIT_WARNING = MAX_NUM_SPRITES - 1;
```
NIT: When I see "SPRITE_LIMIT_WARNING" I tend to think the variable contains a string. Something such as "SPRITE_LIMIT_THRESHOLD" might be a clearer indicator that the variable contains a number.

fisher-alice replied:
Great idea - thanks! I'll use SPRITE_WARNING_THRESHOLD.

jmkulwik commented on apps/src/p5lab/spritelab/CoreLibrary.js:
```diff
@@ -366,12 +369,37 @@ export default class CoreLibrary {
+    if (this.reachedSpriteMax()) {
+      return;
+    }
+    this.warnIfAtSpriteLimit();
```
NIT: (I don't feel strongly about this. Feel free to ignore if you disagree.) Is `reachedSpriteMax` and `warnIfAtSpriteLimit` linked? i.e. you'd generally expect that the two commands should come one after the other? If so, I'd recommend something like this:
```
    if (this.reachedSpriteMax()) {
      return;
    } else if (this.atSpriteLimit()) {
      this.warn();
    }
```
The reasoning: future code maintainers will read an `if` clause followed by a function as two separate logical blocks but will read an `if`/`else if` clause as one logical block. The two are less likely to be separated and it's less likely that one will be deleted while the other retained.

fisher-alice replied:
Thank you! I appreciate your point about future code maintainers and added a helper function similar to reachedSpriteMax called reachedSpriteWarningThreshold. Also renamed the function that dispatches the warning - I think it does make it easier to read with the parallel logic:
    if (this.reachedSpriteMax()) {
      return;
    } else if (this.reachedSpriteWarningThreshold()) {
      this.dispatchSpriteLimitWarning();
    }

jmkulwik reviewed (approved):
Great work on this!!
